src/main/java/net/imglib2/mesh/alg/InteriorPointTest.java

```diff
@@ -0,0 +1,326 @@
+			return false;
+
+		// The rest: ray casting along X.
+		final double ox = p.getDoublePosition( 0 );
+		final double oy = p.getDoublePosition( 1 );
+		final double oz = mround( p.getDoublePosition( 2 ), EPS, 2, 1 );
+
+		// All triangles with minZ < oz
+		int k1 = Arrays.binarySearch( minZs, oz );
+		if ( k1 < 0 )
+			k1 = -( k1 + 1 );
+
+		// Write these indices into an array (so that we do not change
+		// minZs).
+		System.arraycopy( indexMin, 0, indices, 0, k1 );
+
+		/*
+		 * All the triangles with id indices[0], ... indices[jmin-1] (jmin is
+		 * not included, either equal or larger than zr) have a minZ below zr.
+		 *
+		 * Now: search among those triangles what ones have a maxZ larger than
+		 * zr. We will sort indexMin[0 -> jmin] by maxZ.
+		 */
+
+		// Sort indices by the maxZ of the triangle they point to.
+		SortBy.sortBy( indices, maxZs, 0, k1 - 1 );
+
+		// Search the closest z.
+		int k2 = SortBy.binarySearch( indices, maxZs, 0, k1, oz );
+		if ( k2 < 0 )
+			k2 = -( k2 + 1 );
+
+		/*
+		 * All the triangles with id indices[k2], ... indices[k1-1] have a minZ
+		 * below z and a maxZ above z. Only them have to be considered for the
+		 * test.
+		 */
+
+		final int start = k2;
+		final int end = k1;
+		xIntersect.resetQuick();
+		xNormals.resetQuick();
+		for ( int i = start; i < end; i++ )
+		{
+			final int id = indices[ i ];
+			final boolean intersects = rayIntersectsTriangle( id, ox, oy, oz, RX, RY, RZ, intersection );
+			if ( intersects )
+			{
+				xIntersect.add( intersection[ 0 ] );
+				xNormals.add( mesh.triangles().nx( id ) );
+			}
+		}
+
+		if ( xIntersect.size() < 1 )
+			return false;
+
+		/**
+		 * Only consider intersections that are unique. If we have twice the
+		 * same X intersection, then it means we crossed an edge. In that case
+		 * we need to test whether the 2 triangles of this edge are facing the
+		 * same direction. If yes (their normals along x have the same sign), it
+		 * means that we are crossing the mesh boundary. If not, they do not
+		 * count as crossing the boundary.
+		 * 
+		 * Fantastic drawing that illustrates this situation:
+		 * 
+		 * <pre>
+		 *  _________________________________
+		 *  |                               |
+		 *  |                               |
+		 *  |   X --------> .               |
+		 *  |              / \              |
+		 *  |_____________/   \_____________|
+		 * 
+		 * </pre>
+		 * 
+		 * The X point is inside, but the ray-casting will cross the edge
+		 * protruding from the bottom, exactly on the line. In that case this
+		 * should not count as one crossing of the bounds.
+		 */
+		// Sort intersection coords.
+		xIntersect.sort();
+		int nCross = 0;
+
+		double previousX = Double.NaN;
+		double previousN = xNormals.get( 0 );
+		for ( int i = 0; i < xIntersect.size(); i++ )
+		{
+			final double v = xIntersect.get( i );
+			final double n = xNormals.get( i );
+
+			if ( ( v != previousX ) || ( n * previousN > 0 ) )
+				nCross++;
```

Comment:
Could be simplified to `inside = !inside` with `boolean inside` instead of `int nCross`. Then if there are more than 2^31 crossings we won't overflow :sweat_smile: (of course I'm joking, the `xIntersect` array can't get that large anyway)

Reply:
It was useful when debugging. I am sure we will debug again. Let's leave it.